### PR TITLE
python310Packages.paddlepaddle: init at 2.5.0

### PR DESCRIPTION
**Pull request by happysalada**
## Description of changes

this is just a PR to test some fixes on ofborg.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by happysalada** [comment]
@ofborg eval

**Comment by happysalada** [comment]
@ofborg eval python310Packages.paddlepaddle

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+, lib
+, buildPythonPackage
+, fetchPypi
+, python
+, pythonOlder
+, pythonAtLeast
+, openssl_1_1
+, zlib
+, setuptools
+, cudaSupport ? false
```
Comment (natsukium) [inline pkgs/development/python-modules/paddlepaddle/default.nix:11]:
```suggestion
, cudaSupport ? config.cudaSupport or false
```

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+, protobuf
+, pillow
+, decorator
+, astor
+, paddle-bfloat
+, opt-einsum
+}:
+
+let
+  pname = "paddlepaddle${if cudaSupport then "_gpu" else ""}";
```
Comment (natsukium) [inline pkgs/development/python-modules/paddlepaddle/default.nix:25]:
```suggestion
  pname = "paddlepaddle" + lib.optionalString cudaSupport "-gpu";
```

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      cudnn
+    ])
+  );
+
+  postFixup = lib.optionalString stdenv.isLinux ''
+    function fixRunPath {
+      p=$(patchelf --print-rpath $1)
+      patchelf --set-rpath "$p:$libraryPath" $1
+    }
+    fixRunPath $out/lib/python${python.pythonVersion}/site-packages/paddle/fluid/libpaddle.so
```
Comment (natsukium) [inline pkgs/development/python-modules/paddlepaddle/default.nix:60]:
```suggestion
    fixRunPath $out/${python.sitePackages}/paddle/fluid/libpaddle.so
```

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  '';
+
+  buildInputs = [
```
Comment (natsukium) [inline pkgs/development/python-modules/paddlepaddle/default.nix:63]:
```suggestion
  nativeBuildInputs = [
```

**Comment by happysalada** [comment]
@natsukium Thanks for the review.
I've commited those directly, and will just squash those together to retain your attribution when merging.
Let me know if there is anything else of course.

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 5]
```diff
@@ -0,0 +1,90 @@
+{ stdenv
+, lib
```
Comment (natsukium) [inline pkgs/development/python-modules/paddlepaddle/default.nix:2]:
```suggestion
{ stdenv
, config
, lib
```
Sorry, I forgot this.

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 250115 --extra-nixpkgs-config '{ cudaSupport = true; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+let
+  pname = "paddlepaddle" + lib.optionalString cudaSupport "-gpu";
+  version = "2.5.0";
+  format = "wheel";
+  pyShortVersion = "cp${builtins.replaceStrings ["."] [""] python.pythonVersion}";
+  allHashAndPlatform = import ./binary-hashes.nix;
+  hash = allHashAndPlatform."${stdenv.system}"."${if cudaSupport then "gpu" else "cpu"}"."${pyShortVersion}";
+  platform = allHashAndPlatform."${stdenv.system}".platform;
+  src = fetchPypi ({
+    inherit pname version format hash platform;
```
Comment (natsukium) [inline pkgs/development/python-modules/paddlepaddle/default.nix:34]:
```suggestion
    inherit version format hash platform;
    pname = builtins.replaceStrings [ "-" ] [ "_" ] pname;
```

`❯ NIXPKGS_ALLOW_INSECURE=1 nix-build -I nixpkgs=. -E 'with (import <nixpkgs>{}); `
```
these 2 derivations will be built:
  /nix/store/4z47xwvsrjs1rk2g8sb7kkllj9vc97k3-paddlepaddle-gpu-2.5.0-cp310-cp310-manylinux1_x86_64.whl.drv
  /nix/store/1j76p9f22124576hdyvcpg0wapzy6fda-python3.10-paddlepaddle-gpu-2.5.0.drv
building '/nix/store/4z47xwvsrjs1rk2g8sb7kkllj9vc97k3-paddlepaddle-gpu-2.5.0-cp310-cp310-manylinux1_x86_64.whl.drv'...

trying https://files.pythonhosted.org/packages/cp310/p/paddlepaddle-gpu/paddlepaddle-gpu-2.5.0-cp310-cp310-manylinux1_x86_64.whl
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404
error: cannot download paddlepaddle-gpu-2.5.0-cp310-cp310-manylinux1_x86_64.whl from any mirror
error: builder for '/nix/store/4z47xwvsrjs1rk2g8sb7kkllj9vc97k3-paddlepaddle-gpu-2.5.0-cp310-cp310-manylinux1_x86_64.whl.drv' failed with exit code 1;
       last 7 log lines:
       >
       > trying https://files.pythonhosted.org/packages/cp310/p/paddlepaddle-gpu/paddlepaddle-gpu-2.5.0-cp310-cp310-manylinux1_x86_64.whl
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
       > curl: (22) The requested URL returned error: 404
       > error: cannot download paddlepaddle-gpu-2.5.0-cp310-cp310-manylinux1_x86_64.whl from any mirror
       For full logs, run 'nix-store -l /nix/store/4z47xwvsrjs1rk2g8sb7kkllj9vc97k3-paddlepaddle-gpu-2.5.0-cp310-cp310-manylinux1_x86_64.whl.drv'.
error: 1 dependencies of derivation '/nix/store/1j76p9f22124576hdyvcpg0wapzy6fda-python3.10-paddlepaddle-gpu-2.5.0.drv' failed to build
```

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    }
+    fixRunPath $out/${python.sitePackages}/paddle/fluid/libpaddle.so
+  '';
+
+  nativeBuildInputs = [
+    setuptools
+  ];
+
+  propagatedBuildInputs = [
+    httpx
```
Comment (natsukium) [inline pkgs/development/python-modules/paddlepaddle/default.nix:70]:
```suggestion
  propagatedBuildInputs = [
    setuptools
    httpx
```

**Comment by natsukium** [comment]
It seems to recognize my GeForce RTX3080.
```
❯ python -c "import paddle; paddle.utils.run_check()"
Running verify PaddlePaddle program ...
I0819 22:58:56.524948 390677 interpretercore.cc:237] New Executor is Running.
W0819 22:58:56.525087 390677 gpu_resources.cc:119] Please NOTE: device: 0, GPU Compute Capability: 8.6, Driver API Version: 12.1, Runtime API Version: 11.8
W0819 22:58:56.612347 390677 gpu_resources.cc:149] device: 0, cuDNN Version: 8.9.
I0819 22:59:00.330322 390677 interpreter_util.cc:518] Standalone Executor is Used.
PaddlePaddle works well on 1 GPU.
PaddlePaddle is installed successfully! Let's start deep learning with PaddlePaddle now.
```

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+      cudatoolkit.lib
+      cudatoolkit.out
+      cudnn
+    ])
+  );
+
+  postFixup = lib.optionalString stdenv.isLinux ''
+    function fixRunPath {
+      p=$(patchelf --print-rpath $1)
+      patchelf --set-rpath "$p:$libraryPath" $1
```
Comment (natsukium) [inline pkgs/development/python-modules/paddlepaddle/default.nix:60]:
```suggestion
      patchelf --set-rpath "$p:$libraryPath" $1
      ${lib.optionalString cudaSupport ''
        addOpenGLRunpath $1
      ''}
```

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    }
+    fixRunPath $out/${python.sitePackages}/paddle/fluid/libpaddle.so
+  '';
```
Comment (natsukium) [inline pkgs/development/python-modules/paddlepaddle/default.nix:63]:
```suggestion
  '';
  
  nativeBuildInputs = [
    addOpenGLRunpath
  ];
```

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+, buildPythonPackage
+, fetchPypi
+, python
+, pythonOlder
+, pythonAtLeast
+, openssl_1_1
+, zlib
+, setuptools
+, cudaSupport ? config.cudaSupport or false
+, cudaPackages_11 ? {}
```
Comment (natsukium) [inline pkgs/development/python-modules/paddlepaddle/default.nix:13]:
```suggestion
, cudaPackages_11 ? {}
, addOpenGLRunpah
```

**Review by natsukium — COMMENTED** [review]
I think that's all.

File: pkgs/development/python-modules/paddlepaddle/default.nix [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    opt-einsum
+  ];
+
+  pythonImportsCheck = [ "paddle" ];
+
+  # no tests
+  doCheck = false;
+
+  meta = with lib; {
+    description = "PArallel Distributed Deep LEarning: Machine Learning Framework from Industrial Practice （『飞桨』核心框架，深度学习&机器学习高性能单机、分布式训练和跨平台部署";
```
Comment (dotlambda) [inline pkgs/development/python-modules/paddlepaddle/default.nix:91]:
```suggestion
    description = "PArallel Distributed Deep LEarning: Machine Learning Framework from Industrial Practice";
```

**Review by dotlambda — COMMENTED** [review]
`meta.sourceProvenance` missing

**Pull request merged**